homeassistant/components/withings/__init__.py

```diff
@@ -59,52 +82,124 @@ async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
-    if "auth_implementation" not in entry.data:
-        _LOGGER.debug("Upgrading existing config entry")
-        data = entry.data
-        creds = data.get(const.CREDENTIALS, {})
-        hass.config_entries.async_update_entry(
-            entry,
-            data={
-                "auth_implementation": const.DOMAIN,
-                "implementation": const.DOMAIN,
-                "profile": data.get("profile"),
-                "token": {
-                    "access_token": creds.get("access_token"),
-                    "refresh_token": creds.get("refresh_token"),
-                    "expires_at": int(creds.get("token_expiry")),
-                    "type": creds.get("token_type"),
-                    "userid": creds.get("userid") or creds.get("user_id"),
-                },
+    # Normalize the config entry.
+    webhook_id = entry.data.get(CONF_WEBHOOK_ID, webhook.async_generate_id())
+    hass.config_entries.async_update_entry(
+        entry,
+        unique_id=entry.data["token"]["userid"],  # Upgrade existing configs.
+        data={
+            **entry.data,
+            **{
+                const.CLIENT_ID: hass.data[DOMAIN][const.CONFIG][const.CLIENT_ID],
+                const.CLIENT_SECRET: hass.data[DOMAIN][const.CONFIG][
+                    const.CLIENT_SECRET
+                ],
+                const.USE_WEBHOOK: hass.data[DOMAIN][const.CONFIG][const.USE_WEBHOOK],
+                CONF_WEBHOOK_ID: webhook_id,
+                const.CONF_WEBHOOK_URL: entry.data.get(
+                    const.CONF_WEBHOOK_URL, webhook.async_generate_url(hass, webhook_id)
+                ),
             },
-        )
-
-    implementation = await config_entry_oauth2_flow.async_get_config_entry_implementation(
-        hass, entry
+        },
     )
 
-    data_manager = get_data_manager(hass, entry, implementation)
+    data_manager = await async_init_data_manager(hass, entry)
 
-    _LOGGER.debug("Confirming we're authenticated")
+    webhook.async_register(
+        hass,
+        const.DOMAIN,
+        "Withings notify",
+        data_manager.webhook_config.id,
+        async_webhook_handler,
+    )
+
+    _LOGGER.debug("Confirming %s is authenticated to withings", data_manager.profile)
     try:
-        await data_manager.check_authenticated()
+        await data_manager.async_check_authenticated()
     except NotAuthenticatedError:
-        _LOGGER.error(
-            "Withings auth tokens exired for profile %s, remove and re-add the integration",
-            data_manager.profile,
+        # Returning true so async_unload_entry will work later.
+        return True
+
+    async def start_platforms():
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_setup(
+                    entry, BINARY_SENSOR_DOMAIN
+                ),
+                hass.config_entries.async_forward_entry_setup(entry, SENSOR_DOMAIN),
+            ],
+            loop=hass.loop,
```

Comment:
Not needed to pass the loop.

Also don't need to create a list to then send in as positional parameters, just pass in positional parameters :)

```suggestion       
            hass.config_entries.async_forward_entry_setup(
                entry, BINARY_SENSOR_DOMAIN
            ),
            hass.config_entries.async_forward_entry_setup(entry, SENSOR_DOMAIN),
```